pkg/autoscaler/autoscaler.go

```diff
@@ -3,126 +3,208 @@ package autoscaler
-					"resourceName", resourceName,
-					"value", minMetric.value,
-					"deltaSeconds", t.Sub(minMetric.timestamp).Seconds(),
-					"windowSize", as.windowSize)
-			} else {
-				as.logger.Debug("Resource metrics are above threshold")
+			resourceName := item.DescribedObject.Name
+			value := int(item.Value.MilliValue())
+
+			as.logger.DebugWith("Got Metric Entry",
+				"resourceName", resourceName,
+				"metricName", metricName,
+				"value", value)
+
+			if _, found := resourcesMetricsMap[resourceName]; !found {
+				resourcesMetricsMap[resourceName] = make(map[string]int)
 			}
 
-			// rebuild the slice, excluding any old metrics
-			var newMetrics []metricEntry
-			for _, metric := range resourceMetrics {
-				if t.Sub(metric.timestamp) <= as.windowSize {
-					newMetrics = append(newMetrics, metric)
-				}
+			// sanity
+			if _, found := resourcesMetricsMap[resourceName][metricName]; found {
+				return make(map[string]map[string]int), errors.New("Can not have more than one metric value per resource")
 			}
-			as.metricsMap[resourceName][as.metricName] = newMetrics
+
+			resourcesMetricsMap[resourceName][metricName] = value
 		}
 	}
+
+	return resourcesMetricsMap, nil
 }
 
-func (as *Autoscaler) addMetricEntry(resourceName scaler_types.Resource, metricType string, entry metricEntry) {
-	if _, found := as.metricsMap[resourceName]; !found {
-		as.metricsMap[resourceName] = make(map[string][]metricEntry)
+func (as *Autoscaler) checkResourceToScale(resource scaler_types.Resource, resourcesMetricsMap map[string]map[string]int) bool {
+	if _, found := resourcesMetricsMap[resource.Name]; !found {
+		as.logger.DebugWith("Resource does not have metrics data yet, keeping up", "resourceName", resource.Name)
+		return false
+	}
+
+	for _, scaleResource := range resource.ScaleResources {
+		metricName := scaleResource.GetKubernetesMetricName()
+		value, found := resourcesMetricsMap[resource.Name][metricName]
+		if !found {
+			as.logger.DebugWith("One of the metrics is missing data, keeping up",
+				"resourceName", resource.Name,
+				"metricName", metricName)
+			return false
+		}
+
+		if value > scaleResource.Threshold {
+			as.logger.DebugWith("Metric value above threshold, keeping up",
+				"resourceName", resource.Name,
+				"metricName", metricName,
+				"threshold", scaleResource.Threshold,
+				"value", value)
+			return false
+		}
+
+		as.logger.DebugWith("Metric value below threshold",
+			"resourceName", resource.Name,
+			"metricName", metricName,
+			"threshold", scaleResource.Threshold,
+			"value", value)
 	}
-	as.metricsMap[resourceName][metricType] = append(as.metricsMap[resourceName][metricType], entry)
+
+	as.logger.DebugWith("All metric values below threshold, should scale to zero", "resourceName", resource.Name)
+	return true
 }
 
-func (as *Autoscaler) ReportMetric(metric metricEntry) error {
+func (as *Autoscaler) getBiggestWindow(resource scaler_types.Resource) time.Duration {
```

Comment:
1. WindowSize is a bit unclear... Window Size of what? Isn't this the `ScaleEventDebounceDuration`? 
2. `getMax<whatever we decide in (1)>`

Reply:
`getMaxScaleResourceWindowSize`